### PR TITLE
[test] | [eks] change: Create a new EKS cluster for each PR test set

### PR DESCRIPTION
**Pull request by saimidu**
*Issue #, if available:*

## Checklist
- [x] I've prepended PR tag with frameworks/job this applies to : [mxnet, tensorflow, pytorch] | [build] | [test] | [build, test] | [ec2, ecs, eks, sagemaker]
- [x] (If applicable) I've documented below the DLC image/dockerfile this relates to
- [x] (If applicable) I've documented below the tests I've run on the DLC image
- [x] (If applicable) I've reviewed the licenses of updated and new binaries and their dependencies to make sure all licenses are on the Apache Software Foundation Third Party License Policy Category A or Category B license list.  See [https://www.apache.org/legal/resolved.html](https://www.apache.org/legal/resolved.html).
- [x] (If applicable) I've scanned the updated and new binaries to make sure they do not have vulnerabilities associated with them.

*Description:*
- PRs previously used one single central PR-cluster for EKS tests, which had a preconfigured number of nodes. The pods from all PRs were scheduled on this single cluster.
- This design leads to some pods failing to be scheduled when there are a large number of parallel EKS tests.
- This problem has been resolved by creating an EKS cluster for PRs with a smaller number of nodes, so that each PR will have its own set of nodes upon which pods can be scheduled.

*Tests run:*
Tested locally by executing EKS PyTorch tests for all PyTorch images at the same time, because PyTorch EKS tests are where this issue manifests.

*DLC image/dockerfile:*

*Additional context:*

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license. I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: test/testrunner.py [hunk 1]
```diff
@@ -123,16 +123,21 @@ def pull_dlc_images(images):
...
     for long_name, short_name in frameworks.items():
         if long_name in dlc_images:
-            cluster_name = None
-            if not is_pr_context():
-                num_nodes = 3 if long_name != "pytorch" else 4
-                cluster_name = f"dlc-{short_name}-cluster-" \
-                               f"{os.getenv('CODEBUILD_RESOLVED_SOURCE_VERSION')}-{random.randint(1, 10000)}"
-                eks_utils.create_eks_cluster(cluster_name, "gpu", num_nodes, "p3.16xlarge", "pytest.pem")
-                terminable_clusters.append(cluster_name)
+            num_nodes = 1 if is_pr_context() else 3 if long_name != "pytorch" else 4
```
Comment (YYStreet) [inline test/testrunner.py:134]: Will this change make an influence on TF and MXNet EKS tests? Should we also test them under this PR context?

File: test/testrunner.py [hunk 2]
```diff
@@ -123,16 +123,21 @@ def pull_dlc_images(images):
...
+            cluster_name = f"dlc-{short_name}-cluster-" \
+                           f"{os.getenv('CODEBUILD_RESOLVED_SOURCE_VERSION')}-{random.randint(1, 10000)}"
+            eks_utils.create_eks_cluster(cluster_name, "gpu", num_nodes, "p3.16xlarge", "pytest.pem")
+            terminable_clusters.append(cluster_name)
             eks_utils.eks_setup(long_name, cluster_name)
+            break
```
Comment (arjkesh) [inline test/testrunner.py:140]: you can instead just return terminable_clusters here -- but since we are only returning one cluster now, you can just return cluster name and get rid of terminable clusters altogether.

Some logic will need to be changed in main so that eks doesn't expect a list

File: test/testrunner.py (hunk 1 above)
Comment (saimidu) [reply test/testrunner.py]: Yes. It probably will. I'm trying to make it work with PyTorch first, and then resolve issues that may crop up with other frameworks in afterwards.

File: test/testrunner.py (hunk 2 above)
Comment (saimidu) [reply test/testrunner.py:140]: Sure. I will make that change. Thanks.

File: test/testrunner.py [hunk 3]
```diff
@@ -121,19 +121,21 @@ def pull_dlc_images(images):
...
+        raise ValueError(
+            f"All images in dlc_images must be of a single framework for EKS tests.\n"
+            f"Instead seeing {frameworks_in_images} frameworks."
+        )
+    long_name = frameworks_in_images[0]
+    short_name = frameworks[long_name]
+    num_nodes = 2 if is_pr_context() else 3 if long_name != "pytorch" else 4
+    cluster_name = f"dlc-{short_name}-cluster-" \
+                   f"{os.getenv('CODEBUILD_RESOLVED_SOURCE_VERSION')}-{random.randint(1, 10000)}"
+    eks_utils.create_eks_cluster(cluster_name, "gpu", num_nodes, "p3.16xlarge", "pytest.pem")
```
Comment (TusharKanekiDey) [inline test/testrunner.py:136]: Can we use p3.8xlarge for PyTorch ? Previously we used to have that instead.

File: test/testrunner.py [hunk 4]
```diff
@@ -158,15 +160,14 @@ def main():
         if specific_test_type == "sanity":
             pull_dlc_images(all_image_list)
         if specific_test_type == "eks":
-            eks_terminable_clusters = setup_eks_clusters(dlc_images)
+            new_eks_cluster_name = setup_eks_clusters(dlc_images)
```
Comment (arjkesh) [inline test/testrunner.py:163]: nit: let's name this something like "eks_test_cluster" or "eks_cluster_name" -- the "new" is not needed

File: test/testrunner.py [hunk 5]
```diff
@@ -158,15 +160,14 @@ def main():
...
         # Execute dlc_tests pytest command
         pytest_cmd = ["-s", "-rA", test_path, f"--junitxml={report}", "-n=auto"]
         try:
             sys.exit(pytest.main(pytest_cmd))
         finally:
-            if specific_test_type == "eks" and eks_terminable_clusters:
-                for cluster in eks_terminable_clusters:
-                    eks_utils.delete_eks_cluster(cluster)
+            if specific_test_type == "eks":
```
Comment (arjkesh) [inline test/testrunner.py:169]: Can you also add "new_eks_cluster_name" as a condition here? The reason being, why set it to None in the first place -- if it somehow sneaks through as None, we will try to call delete_eks_cluster on "None"

File: test/testrunner.py [hunk 6]
```diff
@@ -121,19 +121,21 @@ def pull_dlc_images(images):
...
+            f"All images in dlc_images must be of a single framework for EKS tests.\n"
+            f"Instead seeing {frameworks_in_images} frameworks."
+        )
+    long_name = frameworks_in_images[0]
+    short_name = frameworks[long_name]
+    num_nodes = 2 if is_pr_context() else 3 if long_name != "pytorch" else 4
+    cluster_name = f"dlc-{short_name}-cluster-" \
+                   f"{os.getenv('CODEBUILD_RESOLVED_SOURCE_VERSION')}-{random.randint(1, 10000)}"
+    eks_utils.create_eks_cluster(cluster_name, "gpu", num_nodes, "p3.16xlarge", "pytest.pem")
+    eks_utils.eks_setup(long_name, cluster_name)
```
Comment (arjkesh) [inline test/testrunner.py:137]: There is a condition i just realized where create could succeed but setup could fail, and we would not properly clean up the cluster. I think it might be wise to wrap the eks setup call in try/finally and always return cluster name after we've created the eks cluster, so that in this case we won't leave clusters spun up
